website/sales/models/shift.py

```diff
@@ -0,0 +1,171 @@
+                        then=Value(True),
+                    ),
+                    default=Value(False),
+                    output_field=BooleanField(),
+                ),
+                total_revenue=Coalesce(
+                    Subquery(
+                        apps.get_model("sales.Order")
+                        .objects.filter(shift=OuterRef("pk"))
+                        .values("shift")
+                        .annotate(_shift_revenue=Sum("total_amount"))
+                        .values("_shift_revenue")
+                    ),
+                    Value(0.00),
+                ),
+                num_orders=Coalesce(
+                    Subquery(
+                        apps.get_model("sales.Order")
+                        .objects.filter(shift=OuterRef("pk"))
+                        .values("shift")
+                        .annotate(cnt=Count("*"))
+                        .values("cnt")
+                    ),
+                    Value(0),
+                ),
+                total_revenue_paid=Coalesce(
+                    Subquery(
+                        apps.get_model("sales.Order")
+                        .objects.exclude(
+                            (Q(payment__isnull=True) & Q(total_amount__gt=0))
+                            | Q(order_items__isnull=True)
+                        )
+                        .filter(shift=OuterRef("pk"))
+                        .values("shift")
+                        .annotate(_shift_revenue=Sum("total_amount"))
+                        .values("_shift_revenue")
+                    ),
+                    Value(0.00),
+                ),
+                num_orders_paid=Coalesce(
+                    Subquery(
+                        apps.get_model("sales.Order")
+                        .objects.exclude(
+                            (Q(payment__isnull=True) & Q(total_amount__gt=0))
+                            | Q(order_items__isnull=True)
+                        )
+                        .filter(shift=OuterRef("pk"))
+                        .values("shift")
+                        .annotate(cnt=Count("*"))
+                        .values("cnt")
+                    ),
+                    Value(0),
+                ),
+            )
+        )
+
+
+class Shift(models.Model):
+
+    objects = ShiftManager()
+
+    start = models.DateTimeField(blank=False, null=False,)
+    end = models.DateTimeField(blank=False, null=False,)
+
+    custom_title = models.CharField(blank=True, null=True, max_length=100)
+
+    product_list = models.ForeignKey(
+        ProductList, blank=False, null=False, on_delete=models.PROTECT
+    )
+
+    locked = models.BooleanField(
+        blank=False,
+        null=False,
+        default=False,
+        help_text=_(
+            "Prevent orders being changed or created for this shift. This will also clean up all unpaid orders in this shift."
+        ),
+    )
```

Comment:
Right now, no not from the admin. If we would ever want to do that, it should only be possible with special permissions (only the board) and even then I think we should log it specifically (because of audit stuff). 

So: no

Reply:
For future reference: logging is possible to write to the django object history.

Reply:
@se-bastiaan what do you mean exactly? I don't see anything specific in the payments app with special object history

Reply:
Oh sorry it is registrations: https://github.com/svthalia/concrexit/blob/master/website/registrations/services.py#L176